app/com/arpnetworking/metrics/portal/alerts/scheduling/AlertExecutionContext.java

```diff
@@ -79,4 +90,153 @@ public AlertExecutionContext(
+                    ImmutableList.of(queryResult)
+            );
+        }
+
+        final long uniqueNames = results.stream()
+                .map(TimeSeriesResult.Result::getName)
+                .distinct()
+                .limit(2)
+                .count();
+
+        final String name = results.get(0).getName();
+        if (uniqueNames > 1) {
+            throw newCompletionException(
+                    PROBLEM_UNEXPECTED_RESULT,
+                    "All result metric names must be identical",
+                    ImmutableList.of(queryResult)
+            );
+        }
+
+        if (results.size() > 1 && results.stream().anyMatch(res -> !getTagGroup(res).isPresent())) {
+            throw newCompletionException(
+                    PROBLEM_UNEXPECTED_RESULT,
+                    "All results must contain a tag group-by if there are multiple results",
+                    ImmutableList.of(queryResult)
+            );
+        }
+
+        final List<Map<String, String>> firingTagGroups =
+                results
+                    .stream()
+                    .filter(res -> isFiring(res.getValues(), period, scheduled))
+                    .map(res -> getTagGroup(res).orElseGet(ImmutableMap::of))
+                    .collect(ImmutableList.toImmutableList());
+
+        return new DefaultAlertEvaluationResult.Builder()
+                .setSeriesName(name)
+                .setFiringTags(firingTagGroups)
+                .build();
+    }
+
+    private boolean isFiring(final List<? extends TimeSeriesResult.DataPoint> values,
+                             final ChronoUnit period,
+                             final Instant scheduled) {
+        if (values.isEmpty()) {
+            return false;
+        }
+        // The most recent datapoint and scheduled time must belong to the same
+        // period.
+        final Instant mostRecentDatapointTime = values.get(values.size() - 1).getTime();
+        return period.between(mostRecentDatapointTime, scheduled) == 0;
+    }
+
+    private Optional<Map<String, String>> getTagGroup(final TimeSeriesResult.Result result) {
+        // We don't expect the tag group-by to be the only value in the stream
+        // because the histogram plugin uses its own group-by.
+        return result.getGroupBy().stream()
+                .filter(g -> g instanceof TimeSeriesResult.QueryTagGroupBy)
+                .map(g -> (TimeSeriesResult.QueryTagGroupBy) g)
+                .findFirst()
+                .map(TimeSeriesResult.QueryTagGroupBy::getGroup);
+    }
+
+    private BoundedMetricsQuery applyTimeRange(final MetricsQuery query,
+                                               final Instant scheduled,
+                                               final ChronoUnit period) {
+        // We must truncate to avoid dealing with partially aggregated periods.
+        final ZonedDateTime latest = ZonedDateTime.ofInstant(scheduled, ZoneOffset.UTC).truncatedTo(period);
```

Comment:
We want this end-time to be _exclusive_, right? IIRC, KairosDB's semantics are that `end_time` is _inclusive_, so I think we want to subtract 1ns or 1ms or something from this value. We could do that here, if we want the alerting code to have the convention that endTime is inclusive; or [here](https://github.com/ArpNetworking/metrics-portal/blob/6fe9fe26d2cfced47174ba05095a699674ee3734/app/com/arpnetworking/metrics/portal/query/impl/KairosDbQueryExecutor.java#L113), if we want the convention that it's exclusive.

Reply:
This range should be inclusive like it is with the generic query executor. I don't think there's any change necessary if kept inclusive. Subtracting some duration would be essentially treating the range as exclusive, no?

Reply:
Hrrm, but-- suppose
```
period = MINUTELY
scheduled = 2020-06-24 12:00:00
```
Then
```
latest   = 2020-06-24 12:00:00
previous = 2020-06-24 11:59:00
```
...and then we'll fire a request at Kairos like
```json
{ "...": "..." , "start_absolute": "2020-06-24 11:59:00", "end_absolute": "2020-06-24 12:00:00" }
```
...and then we'll get the datapoints for both 11:59:00 _and_ 12:00:00.
Is that what we want?

Reply:
Yes, I think that's alright. The only timestamp we're concerned with when interpreting the result is the one belonging to the most recent datapoint. So if we got both points back, I would only look at the one corresponding to `2020-06-24 12:00:00`. It's also why I then check that the difference between scheduled and the last timestamp is within one period, so for example if the only datapoint was at `11:59:00` then it wouldn't be considered firing at this moment since that was >=1 period ago.

Reply:
All this talk however does make me realize I need a mechanism to account for data delay, because if I'm always looking for a datapoint at the current minute it won't be present due to the 3 minute delay from aggregation and so we'd always report the alert as not firing.

Reply:
Ahhh, indeed. Do you plan to put that in this diff or fast-follow it?

Reply:
I put that in this diff. It was only ~80 lines in total...although it did push this diff over 1k 😬 

https://github.com/ArpNetworking/metrics-portal/pull/422/commits/ce8ceb13aa152e774853c4d4fd674bc584ce5c72